### PR TITLE
feat: add golang purls to PkgInfo [OSM-607]

### PR DESCRIPTION
**Pull request by mcombuechen**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?

This introduces the generation of PackageURLs when a `golang` dependency graph is being constructed. This will enable us to output valid `golang` purls during the export of a project’s SBOM, or the generation of a document on the client side.

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -726,3 +729,31 @@ function toSnykVersion(v: ModuleVersion): string {
     throw new Error('Unexpected module version format');
   }
 }
+
+function constructPurl(mod: GoModule, importPath: string): string | undefined {
+  let namespace: string | undefined;
+  let name: string | undefined;
+  const version = mod.Version;
+  let subpath: undefined | string;
```
Comment (dotkas) [inline lib/index.ts:737]:
```
  let namespace: string | undefined;
  let name: string | undefined;
```
Then
```
let subpath: undefined | string;
```

**Comment by darscan** [comment]
I would hold off on this until we've updated the dep-graph lib across the entire Snyk codebase

File: lib/index.ts [hunk 2]
```diff
@@ -726,3 +729,31 @@ function toSnykVersion(v: ModuleVersion): string {
...
+
+  const idx = mod.Path.lastIndexOf('/');
+  if (idx < 0) {
+    name = mod.Path;
+  } else {
+    namespace = mod.Path.slice(0, idx);
+    name = mod.Path.slice(idx + 1);
+  }
```
Comment (dotkas) [inline lib/index.ts:745]: A comment or two could be nice. I was thinking like an example with actual string and what happens.

File: lib/index.ts [hunk 3]
```diff
@@ -726,3 +729,31 @@ function toSnykVersion(v: ModuleVersion): string {
...
+
+  if (importPath !== mod.Path) {
+    subpath = importPath.replace(`${mod.Path}/`, '');
+  }
```
Comment (dotkas) [inline lib/index.ts:749]: Same.

File: lib/index.ts (hunk 1 above)
Comment (dotkas) [reply lib/index.ts:737]: On a serious note - do they really all have to be potentially `undefined`? 😞

**Review by dotkas — APPROVED** [review]
I'm a fan. Thanks A LOT for doing this for us.

**Review by darscan — CHANGES_REQUESTED** [review]
We can't merge this until it's safe to do so (when a newer dep-graph lib has been rolled out across Snyk)

File: lib/index.ts (hunk 1 above)
Comment (mcombuechen) [reply lib/index.ts:737]: can refactor a bit here, but yeah they are all optional except for `name`. `string | undefined` is an improvement from [`string | undefined | null`](https://github.com/package-url/packageurl-js/blob/3e0e10616b23ba5c1a9529f31357abd5ee0bfa72/src/package-url.d.ts#L63) imho 😆

**Comment by mcombuechen** [comment]
converted this back to a draft. @darscan is right, we need to get the added purl validation logic into several other places first.

**Pull request closed without merging**